### ISSUE TITLE
CS4014 warning when using RecurringJob.AddOrUpdate method without specifying Id

### ISSUE THREAD

**Issue by amattie** [HangfireIO/Hangfire#646]
_Hangfire.Core 1.6.2_

I'm trying to schedule a recurring job like this:

`RecurringJob.AddOrUpdate<MyClass>((_) => _.Run().Wait(), Cron.MinuteInterval(1));`

My MyClass.Run method has this signature:

`public async Task Run() { }`

Hangfire throws an ArgumentException during runtime due to Job's Validate method. This is the specific culprit: https://github.com/HangfireIO/Hangfire/blob/4dec6435fb552364833c359079230ad04fd98e53/src/Hangfire.Core/Common/Job.cs#L384.

Is Hangfire being over-protective here? Even if I can't run fully async, it feels like I should be able to call .Wait() on a Task that's returned from a method on MyClass.

I believe this issue is different than #150 and the discussion on https://discuss.hangfire.io/t/async-task-jobs/73.

**Comment by odinserj** [comment]
Have you tried to specify just call to the `Run` method?

```csharp
RecurringJob.AddOrUpdate<MyClass>((_) => _.Run(), Cron.MinuteInterval(1));
```

**Comment by odinserj** [comment]
Please also keep in mind, that `x => x.Run()` is not a delegate, it is an expression tree.

**Comment by amattie** [comment]
Interesting, that works. I didn't realize that argument was an expression tree, and it doesn't help that Roslyn doesn't explicitly realize it's an expression tree either and thus triggers a [CS4014 warning](https://msdn.microsoft.com/en-us/library/hh873131.aspx?f=255&MSPPError=-2147217396) during build. The Visual Studio UI also highlights it as a potential problem due to that warning.

It still feels like Hangfire is being overly protective here, but I also don't know the consequences of treating this argument as a delegate and allowing `.Wait()` to be called. I'm leaving this open due to the build warning this causes, but feel free to close if you disagree.

**Comment by gilesbradshaw** [comment]
am I understanding this right?  If I have 

````
public static void Go()
		{
			RecurringJob.AddOrUpdate(()=> Process(), Cron.Minutely);

		}
		public static async Task Process()
		{ 
                     await Task.Delay();
                }
````

then hangfire will run process async and wait for its completion?

ie I can ignore the VS green squiggly line?

**Comment by odinserj** [comment]
@amattie, background jobs in Hangfire may span the process boundaries, and all the information about the method invocation should be serialized. There is no simple and efficient way to serialize delegates, because this will require scope serialization as well, that's why expression trees are used. Hangfire serializes only *assembly-qualified type name*, *method name* and *arguments* to restore them during the processing.

However, CS4014 warning is bad and confusing, you are correct. I've added missing overloads for the `AddOrUpdate` method to get rid of it. The fix will be released with Hangfire 1.6.3.

@gilesbradshaw, yes, since version 1.6.0, Hangfire will wait for `async Task` method completion, please see http://hangfire.io/blog/2016/07/16/hangfire-1.6.0.html#async-methods-for-background-jobs
